File: Java/57. Insert Interval.java
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        LinkedList<int[]> res = new LinkedList();
        int start = newInterval[0], end= newInterval[1], i = 0, N = intervals.length;
        
        /* add all old intervals that are before newInterval */
        while (i < N && start > intervals[i][0]) {
            res.add(intervals[i++]);
        }
        
        /* add newInterval to res, no overlap */
        if (res.isEmpty() || res.getLast()[1] < start) {
            res.add(newInterval);
        } 
        /* overlap */
        else {
            newInterval = res.pollLast();
            newInterval[1] = Math.max(newInterval[1], end);
            res.add(newInterval);
        }
        
        /* add all old remaining intervals */
        while (i < N) {
            start = intervals[i][0];
            end   = intervals[i++][1];
            /* no overlap */
            if (res.getLast()[1] < start) {
                res.add(intervals[i-1]);
            } else {
                newInterval = res.pollLast();
                newInterval[1] = Math.max(newInterval[1], end);
                res.add(newInterval);
            }
        }
        
        return res.toArray(new int[res.size()][2]);
    }
}

File: Python3/57. Insert Interval.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        ans = []
        start, end = newInterval
        i, N = 0, len(intervals)
        
        # step 1: add all old intervals before newInterval
        while i < N and start > intervals[i][0]:
            ans.append(intervals[i])
            i += 1
            
        # step 2: add the newInterval
        #         case 1: no overlap
        if not ans or ans[-1][1] < start:
            ans.append(newInterval)
            
        #         case 2: overlap
        else:
            ans[-1][1] = max(ans[-1][1], end)
            
        # step 3: add the remaining old intervals
        while i < N:
            start, end = intervals[i]
            i += 1
            
            # case 1: no overlap
            if ans[-1][1] < start:
                ans.append([start, end])
            # case 2: overlap
            else:
                ans[-1][1] = max(ans[-1][1], end)
            
        return ans